daemon/SecretsImpl/dataprotector.cpp

```diff
@@ -0,0 +1,326 @@
+        bool isOpen = file->open(QIODevice::WriteOnly);
+        if (!isOpen) {
+            qCWarning(lcSailfishSecretsDaemon) << "Can't open file for writing:" << file->fileName();
+        }
+        return isOpen;
+    });
+
+    if (!allOpen) {
+        qCWarning(lcSailfishSecretsDaemon) << "Not all files could be open for writing.";
+        return ErrorCannotOpenFile;
+    }
+
+    for (QSharedPointer<QSaveFile> &file : files) {
+        qint64 bytesWritten = file->write(bytes);
+        if (bytesWritten != bytes.size()) {
+            qCWarning(lcSailfishSecretsDaemon) << "Can't write file:" << file->fileName();
+            return ErrorCannotWriteFile;
+        }
+        file->commit();
+    }
+
+    // Remove old data directories
+    for (QFileInfo &fileInfo : oldDirectories) {
+        QDir oldDataDir(fileInfo.absoluteFilePath());
+        oldDataDir.removeRecursively();
+    }
+
+    // Clear in-memory data, so it gets refreshed on next call
+    m_data.clear();
+    return Success;
+}
+
+
+DataProtector::Status DataProtector::processLegacy()
+{
+    int slashIndex = m_path.lastIndexOf('/');
+    if (slashIndex <= 0) {
+        return ErrorInvalidPath;
+    }
+
+    QString parentPath = m_path.mid(0, slashIndex);
+    QDir parentDir(parentPath);
+    if (!parentDir.exists()) {
+        // The parent directory doesn't even exist, so there is no legacy data in there
+        return Success;
+    }
+
+    // See if m_path is a file, if yes, then we are dealing with legacy data here
+    QFileInfoList files = parentDir.entryInfoList(QDir::Files);
+    auto found = std::find_if(files.begin(), files.end(), [this](const QFileInfo &f) { return f.absoluteFilePath() == m_path; });
+    QString legacyPath = m_path + QStringLiteral("_legacy");
+    if (found == files.end()) {
+        // Could be that the file is already renamed but the device was shut down during writing the new contents
+        found = std::find_if(files.begin(), files.end(), [&legacyPath](const QFileInfo &f) { return f.absoluteFilePath() == legacyPath; });
+
+        if (found == files.end()) {
+            // We determined that m_path is not a file, and the temporary file with _legacy suffix isn't present,
+            // therefore no legacy processing is needed.
+            return Success;
+        }
+    }
+
+    // Rename the file temporarily to allow creating the directory at m_path
+    QFile legacyFile(found->absoluteFilePath());
+    if (found->absoluteFilePath() != legacyPath && !legacyFile.rename(legacyPath)) {
+        qCWarning(lcSailfishSecretsDaemon) << "could not rename file:" << found->absoluteFilePath() << "to:" << legacyPath;
+        return ErrorCannotRenameFile;
+    }
+
+    // Read all data from the file
+    legacyFile.open(QIODevice::ReadOnly);
+    QByteArray fileContents = legacyFile.readAll();
+
+    // Store the data in a redundant manner
+    putData(fileContents);
```

Comment:
putData() may return an error here, no? Then the legacy file should not be removed.